include/up-cpp/utils/SafeMap.h

```diff
@@ -0,0 +1,323 @@
+///
+/// Used to lock the "other" object when copying/moving. Requires an explicit call
+/// to unlock() after the initializer list is done (in the ctor body).
+struct CopyMoveCtorLocker {
+    CopyMoveCtorLocker() = default;
+
+    enum class LockMode {
+        None, Unique, Shared
+    };
+
+    CopyMoveCtorLocker(LockMode mode, std::shared_mutex &m) noexcept;
+
+    LockMode unlock() noexcept;
+
+private:
+    LockMode cmclock_mode{LockMode::None};
+    std::unique_lock<std::shared_mutex> ulock;
+    std::shared_lock<std::shared_mutex> slock;
+};
+} // namespace detail
+
+/// @brief Wraps std::map in a std::shared_mutex for thread-safe access
+///
+/// By using std::shared_mutex, we can avoid serialization of const access to
+/// the map. While there is still _some_ performance hit from adding the lock,
+/// this mitigates the worst of it.
+///
+/// Aside from the locking, all interfaces are passthroughs to the underlying
+/// std::map class.
+///
+/// @remarks We use protected inheritance from std::map so that any attempt to
+///          use an interface that has not been protected by a mutex will fail
+///          to compile with an error that the method is protected.
+///
+/// @remarks On the subject of iterators and transactions:<br>
+///          This wrapper *does not* allow direct access to interfaces that
+///          take or return iterators. This is because iterators already can
+///          be easily invalidated by many operators, so adding concurrent
+///          access makes it nearly impossible to use them. While locking could
+///          be added to the iterators themselves, that would introduce new
+///          risks (e.g. locks accidentally being held because an iterator
+///          was held).
+///          <br>
+///          The solution is to provide transact() methods that allow for bulk
+///          actions while holding the lock. For operations where iterators
+///          or atomic operations need to occur on map data, a callable
+///          (function pointer, lambda, etc) can be passed to transact(). The
+///          lock will be held while the callable is running.
+template<typename MapT>
+class BaseSafeMap : private detail::CopyMoveCtorLocker, protected MapT {
+    mutable std::shared_mutex map_lock_{};
+    using UnsafeMapT = MapT;
+    using SafeMapT = BaseSafeMap<MapT>;
+    using CMCLock = detail::CopyMoveCtorLocker;
+
+public:
+    ///////////////////////////////////////////////////////////////////////////
+    /////////////// New interfaces for safe bulk transactions /////////////////
+    ///////////////////////////////////////////////////////////////////////////
+    using ModifyTxn = std::function<std::any(MapT&)>;
+    std::any transact(ModifyTxn &&f) {
+        std::unique_lock lock(map_lock_);
+        return f(*this);
+    }
+
+    using ConstTxn = std::function<std::any(MapT const&)>;
+    std::any transact(ConstTxn &&f) const {
```

Comment:
Having thought on this a little more, it should be trivial to remove std::any and just use a templated return type. That will also make it easier to perform non-returning transactions.

Reply:
`std::any` has been removed in the latest commit